zerver/lib/scim.py

```diff
@@ -0,0 +1,314 @@
+        self.subdomain = get_subdomain(request)
+        self.config = settings.SCIM_CONFIG[self.subdomain]
+
+        self._email_new_value: Optional[str] = None
+        self._is_active_new_value: Optional[bool] = None
+        self._full_name_new_value: Optional[str] = None
+        self._password_set_to: Optional[str] = None
+
+    def is_new_user(self) -> bool:
+        return not bool(self.obj.id)
+
+    @property
+    def display_name(self) -> str:
+        """
+        Return the displayName of the user per the SCIM spec.
+
+        Overriden because UserProfile uses the .full_name attribute,
+        while the superclass expects .first_name and .last_name.
+        """
+        return self.obj.full_name
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Return a ``dict`` conforming to the SCIM User Schema,
+        ready for conversion to a JSON object.
+        """
+        if self.config["name_formatted_included"]:
+            name = {
+                "formatted": self.obj.full_name,
+            }
+        else:
+            if " " not in self.obj.full_name:
+                first_name, last_name = "", self.obj.full_name
+            else:
+                first_name, last_name = self.obj.full_name.split(" ", 1)
+            name = {
+                "givenName": first_name,
+                "familyName": last_name,
+            }
+        d = dict(
+            {
+                "schemas": [scim_constants.SchemaURI.USER],
+                "id": self.obj.id,
+                "userName": self.obj.delivery_email,
+                "name": name,
+                "displayName": self.display_name,
+                "active": self.obj.is_active,
+                # meta is a property implemented in the superclass
+                "meta": self.meta,
+            }
+        )
+
+        return d
+
+    def from_dict(self, d: Dict[str, Any]) -> None:
+        """
+        Consume a ``dict`` conforming to the SCIM User Schema. The dict is originally submitted
+        as JSON by the client in PUT (update a user) and POST (create a new user) requests.
+        A PUT request tells us to update User attributes to match those passed in the dict.
+        A POST request tells us to create a new User with attributes as specified in the dict.
+
+        Completely overriden because we handle things differently than the superclass. We just
+        store the values of the supported attributes before actually proceeding with making changes
+        in self.save().
+        """
+        email = d.get("userName")
+        assert isinstance(email, str)
+        self.validate_email(email)
+        if self.obj.delivery_email != email:
+            self._email_new_value = email
+
+        name_attr_dict = d.get("name", {})
+        if self.config["name_formatted_included"]:
+            full_name = name_attr_dict.get("formatted", "")
+        else:
+            # Some providers (e.g. Okta) don't provide name.formatted.
```

Comment:
Same here, I wish this block wasn't needed, but Okta doesn't tell us what the full name is